src/traces/funnelarea/plot.js

```diff
@@ -0,0 +1,306 @@
+
+var Drawing = require('../../components/drawing');
+var Lib = require('../../lib');
+var svgTextUtils = require('../../lib/svg_text_utils');
+
+var barPlot = require('../bar/plot');
+var getTransformToMoveInsideBar = barPlot.getTransformToMoveInsideBar;
+
+var pieHelpers = require('../pie/helpers');
+var piePlot = require('../pie/plot');
+
+var attachFxHandlers = piePlot.attachFxHandlers;
+var determineInsideTextFont = piePlot.determineInsideTextFont;
+
+var scalePies = piePlot.scalePies;
+var prerenderTitles = piePlot.prerenderTitles;
+var positionTitleOutside = piePlot.positionTitleOutside;
+
+module.exports = function plot(gd, cdModule) {
+    var fullLayout = gd._fullLayout;
+
+    prerenderTitles(cdModule, gd);
+    scalePies(cdModule, fullLayout._size);
+
+    Lib.makeTraceGroups(fullLayout._funnelarealayer, cdModule, 'trace').each(function(cd) {
+        var plotGroup = d3.select(this);
+        var cd0 = cd[0];
+        var trace = cd0.trace;
+
+        setCoords(cd);
+
+        // TODO: miter might look better but can sometimes cause problems
+        // maybe miter with a small-ish stroke-miterlimit?
+        plotGroup.attr('stroke-linejoin', 'round');
+
+        plotGroup.each(function() {
+            var slices = d3.select(this).selectAll('g.slice').data(cd);
+
+            slices.enter().append('g')
+                .classed('slice', true);
+            slices.exit().remove();
+
+            var quadrants = [
+                [[], []], // y<0: x<0, x>=0
+                [[], []] // y>=0: x<0, x>=0
+            ];
+
+            slices.each(function(pt) {
+                if(pt.hidden) {
+                    d3.select(this).selectAll('path,g').remove();
+                    return;
+                }
+
+                // to have consistent event data compared to other traces
+                pt.pointNumber = pt.i;
+                pt.curveNumber = trace.index;
+
+                quadrants[pt.pxmid[1] < 0 ? 0 : 1][pt.pxmid[0] < 0 ? 0 : 1].push(pt);
+
+                var cx = cd0.cx;
+                var cy = cd0.cy;
+                var sliceTop = d3.select(this);
+                var slicePath = sliceTop.selectAll('path.surface').data([pt]);
+
+                slicePath.enter().append('path')
+                    .classed('surface', true)
+                    .style({'pointer-events': 'all'});
+
+                sliceTop.call(attachFxHandlers, gd, cd);
+
+                var shape =
+                    'M' + cx + ',' + cy +
+                    move(pt.TR) +
+                    line(pt.TR, pt.BR) +
+                    line(pt.BR, pt.BL) +
+                    line(pt.BL, pt.TL) +
+                    'Z';
+
+                slicePath.attr('d', shape);
+
+                // add text
+                var textPosition = pieHelpers.castOption(trace.textposition, pt.pts);
+                var sliceTextGroup = sliceTop.selectAll('g.slicetext')
+                    .data(pt.text && (textPosition !== 'none') ? [0] : []);
```

Comment:
This works. But currently for `funnelarea`, `textposition` isn't an available attributes. So these lines here are equivalent to: 

```js
var sliceTextGroup = sliceTop.selectAll('g.slicetext'.data(pt.text ? [0] : []);
```
---

That said, if we're planning to add support for `textposition: 'outside'` at some points, it might best to add `textposition` in this PR as an enumerated with values `['inside', 'none']`.

Reply:
Nicely done in https://github.com/plotly/plotly.js/pull/3876/commits/8af4d12318381574bf280a88ec348d7c0a8b79b2